### DISCUSSION TITLE
如何使用Cloudflare中转Alist的流量？

### DISCUSSION THREAD

**Discussion by Arc-2023** [AlistGo/alist#2697] [Q&A] [answered]
问题始于阿里云盘多用户下载导致403的问题和阿里云小鸡5M带宽根本不够用
于是本着多用户通过一个Cloudflare服务器中转流量的想法，尝试部署了一下
首先是部署Workers，将Alist给出的最新JS代码
`https://github.com/alist-org/alist-proxy/blob/main/alist-proxy.js`
拷贝出，然后把Host和Token替换成自己的IP以及token

+ 之后是第一个问题：workers的域名在国内是访问不到的
也就是说们直接访问这个是访问不到的
于是我尝试用cloudflare代理一个域名
首先把我的一个已经用Cloudflare代理的域名创建CNAME记录，将down子域名指向workers域名(好像这里填什么都影响不大，因为流量总是会被中转到workers)
然后在cloudflare管理域名界面为子域名添加了Worker Route，如图：
![image](https://user-images.githubusercontent.com/64178177/207241812-1c6b2d08-ffec-4748-a0e9-ad64baecdab5.png)
![image](https://user-images.githubusercontent.com/64178177/207243511-9919dabc-f547-40c6-8c91-7f0e8fa6875e.png)
（那个aa的服务就是alist-proxy的js代码）
此时在国内访问那个down子域名就会返回如下的Json：
`{
"code": 401,
"message": "expire missing"
}`
**也就是说GET请求成功转发到了workers，那个CNAME记录的添加是有效的**（大概吧）

然后我把down子域名写到了alist网盘设置的下载代理域名中，在浏览器网页点击链接并被IDM捕获之后发现下载链接的前缀从我自己的阿里云服务器域名变成了CNAME记录所写的down子域名

+ 然后这就有了第二个问题：无法下载。
首先，IDM捕获到下载请求之后点击下载就会报500(Internal Server Error)或者302，如图：
![image](https://user-images.githubusercontent.com/64178177/207242583-6f46df25-2030-4078-943d-815ba19fbc90.png)

而用Edge下载则会根据302跳转信息跳转到那个下载页面并显示此信息：
![image](https://user-images.githubusercontent.com/64178177/207242930-767db06e-943e-440c-aa06-3716c4cadee3.png)
此时在cloudflare中查看流日志可以看到报错信息：
`{
  "outcome": "exception",
  "scriptName": "aa",
  "exceptions": [
    {
      "name": "SyntaxError",
      "message": "Unexpected token 'e', \"error code: 1003\" is not valid JSON",
      "timestamp": 1670913067165
    }
  ],
  "logs": [],
  "eventTimestamp": 1670913067160,
  "event": {
    "request": {
      "url": "https://down.ghngtfnjufgjnghnjfhnjfghjghjhfgjndfthdtrbhdtrh.tk/ali/MOV/%E8%80%81%E5%8F%8B%E8%AE%B0%E5%85%A8%E5%8D%81%E5%AD%A3.Friends.The.Complete.Series.1994-2003.BluRay.1080p.x265.10bit.MNHD-FRDS/%E8%80%81%E5%8F%8B%E8%AE%B0%E7%AC%AC%E5%85%AB%E5%AD%A3.Friends.S08.2001.BluRay.1080p.x265.10bit.MNHD-FRDS/Friends.S08E10.2001.BluRay.1080p.x265.10bit.MNHD-FRDS.mkv?sign=REDACTED=:0",
      "method": "GET",
      "headers": {
        "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
        "accept-encoding": "gzip",
        "accept-language": "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
        "cf-connecting-ip": "20.212.107.98",
        "cf-ipcountry": "SG",
        "cf-ray": "778ca7ad7f0e5627",
        "cf-visitor": "{\"scheme\":\"https\"}",
        "connection": "Keep-Alive",
        "host": "down.ghngtfnjufgjnghnjfhnjfghjghjhfgjndfthdtrbhdtrh.tk",
        "sec-ch-ua": "\"Not?A_Brand\";v=\"8\", \"Chromium\";v=\"108\", \"Microsoft Edge\";v=\"108\"",
        "sec-ch-ua-mobile": "?0",
        "sec-ch-ua-platform": "\"Windows\"",
        "sec-fetch-dest": "document",
        "sec-fetch-mode": "navigate",
        "sec-fetch-site": "cross-site",
        "sec-fetch-user": "?1",
        "upgrade-insecure-requests": "1",
        "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36 Edg/108.0.1462.46",
        "x-forwarded-proto": "https",
        "x-real-ip": "20.212.107.98"
      },
      "cf": {
        "clientTcpRtt": 2,
        "longitude": "103.85540",
        "latitude": "1.30360",
        "tlsCipher": "AEAD-AES128-GCM-SHA256",
        "continent": "AS",
        "asn": 8075,
        "clientAcceptEncoding": "gzip, deflate, br",
        "country": "SG",
        "tlsClientAuth": {
          "certIssuerDNLegacy": "",
          "certIssuerSKI": "",
          "certSubjectDNRFC2253": "",
          "certSubjectDNLegacy": "",
          "certFingerprintSHA256": "",
          "certNotBefore": "",
          "certSKI": "",
          "certSerial": "",
          "certIssuerDN": "",
          "certVerified": "NONE",
          "certNotAfter": "",
          "certSubjectDN": "",
          "certPresented": "0",
          "certRevoked": "0",
          "certIssuerSerial": "",
          "certIssuerDNRFC2253": "",
          "certFingerprintSHA1": ""
        },
        "tlsExportedAuthenticator": {
          "clientFinished": "395f2eeddc1c983ec34593ca5e36af1a171fcc3dc9de893bc4ac17dc745e6992",
          "clientHandshake": "0161a4f6c2c6c3d196d14b58ec17aeb897b643edf894c9722f4c87c0b9f6497f",
          "serverHandshake": "1ba578616d5b00e19664c1be6b19c782769423c7ecedfaa6d92518a86ce9221c",
          "serverFinished": "1489494780bbca0964b717cecec74e07020066c6d58833ae73d93ef9f5c13030"
        },
        "tlsVersion": "TLSv1.3",
        "colo": "SIN",
        "timezone": "Asia/Singapore",
        "edgeRequestKeepAliveStatus": 1,
        "requestPriority": "weight=256;exclusive=1",
        "city": "Singapore",
        "httpProtocol": "HTTP/2",
        "asOrganization": "Microsoft Azure",
        "postalCode": "18"
      }
    },
    "response": {
      "status": 500
    }
  },
  "id": 0
}`
看起来像是我alist的下载请求中的数据没有成功转发到workers导致的报错？这是因为我哪里的操作有问题吗？
请问各路大神，这种问题我该如何解决？

**Comment by Scarecrow-ai** [comment]
遇到了相同的问题，期待解决方法

**Comment by xhofe** [comment]
https://developers.cloudflare.com/fundamentals/get-started/reference/network-ports/

**Reply by Arc-2023** [reply]
将使用docker部署的alist重新部署，端口映射换到80:5244之后改了下cloudflare worker的代码：
![image](https://user-images.githubusercontent.com/64178177/211302803-7981d1f1-174a-41ad-aa98-f1c265bea9bb.png)

使用302转发下载的话,idm显示的下载地址是含beijing的域名:```https://cn-beijing-data.aliyundrive.net/*```
cloudflare应该也是会走这个https域名
所以
+ worker会连接到我的服务器，走的是80端口
+ worker会连接到阿里云的服务器，走的是443端口

这样应该就不会被cloudflare因端口限制连接了吧
**但是还是故障依旧**
![image](https://user-images.githubusercontent.com/64178177/211302633-c07c90b1-0695-45e3-ad16-4f9c7ca75342.png)

Code:
`{
  "outcome": "exception",
  "scriptName": "aa",
  "exceptions": [
    {
      "name": "SyntaxError",
      "message": "Unexpected token 'e', \"error code: 1003\" is not valid JSON",
      "timestamp": 1673265020393
    }
  ],
  "logs": [],
  "eventTimestamp": 1673265020375,
  "event": {
    "request": {
      "url": "http://down.ghngtfnjufgjnghnjfhnjfghjghjhfgjndfthdtrbhdtrh.tk/ali/MOV/God%20of%20War%202023.01.01%20-%2023.14.07.01.mp4?sign=REDACTED=:0",
      "method": "GET",
      "headers": {
        "accept": "*/*",
        "accept-encoding": "gzip",
        "accept-language": "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
        "cf-connecting-ip": "175.41.167.44",
        "cf-ipcountry": "SG",
        "cf-ray": "786cf4690d224733",
        "cf-visitor": "{\"scheme\":\"http\"}",
        "connection": "Keep-Alive",
        "host": "down.ghngtfnjufgjnghnjfhnjfghjghjhfgjndfthdtrbhdtrh.tk",
        "range": "bytes=0-",
        "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36 Edg/108.0.1462.76",
        "x-forwarded-proto": "http",
        "x-real-ip": "175.41.167.44"
      },
      "cf": {
        "clientTcpRtt": 1,
        "longitude": "103.85470",
        "latitude": "1.29290",
        "tlsCipher": "",
        "continent": "AS",
        "asn": 16509,
        "clientAcceptEncoding": "identity;q=1, *;q=0",
        "country": "SG",
        "tlsClientAuth": {
          "certIssuerDNLegacy": "",
          "certIssuerSKI": "",
          "certSubjectDNRFC2253": "",
          "certSubjectDNLegacy": "",
          "certFingerprintSHA256": "",
          "certNotBefore": "",
          "certSKI": "",
          "certSerial": "",
          "certIssuerDN": "",
          "certVerified": "NONE",
          "certNotAfter": "",
          "certSubjectDN": "",
          "certPresented": "0",
          "certRevoked": "0",
          "certIssuerSerial": "",
          "certIssuerDNRFC2253": "",
          "certFingerprintSHA1": ""
        },
        "tlsVersion": "",
        "colo": "SIN",
        "timezone": "Asia/Singapore",
        "edgeRequestKeepAliveStatus": 1,
        "requestPriority": "",
        "asOrganization": "Amazon.com",
        "httpProtocol": "HTTP/1.1"
      }
    },
    "response": {
      "status": 500
    }
  },
  "id": 0
}`

**Comment by m3terpreter** [comment]
您的邮件已收到，谢谢。

**Comment by wonka80** [comment]
收到您的邮件，谢谢！

**Comment by chouqiao** [comment]
> 问题始于阿里云盘多用户下载导致403的问题和阿里云小鸡5M带宽根本不够用 于是本着多用户通过一个Cloudflare服务器中转流量的想法，尝试部署了一下 首先是部署Workers，将Alist给出的最新JS代码 `https://github.com/alist-org/alist-proxy/blob/main/alist-proxy.js` 拷贝出，然后把Host和Token替换成自己的IP`http://119.23.243.88:5244`以及token
> 
>     * 之后是第一个问题：workers的域名在国内是访问不到的`https://aa.a13930228843.workers.dev`
>       也就是说们直接访问这个是访问不到的
>       于是我尝试用cloudflare代理一个域名
>       首先把我的一个已经用Cloudflare代理的域名创建CNAME记录，将down子域名指向workers域名(好像这里填什么都影响不大，因为流量总是会被中转到workers)
>       然后在cloudflare管理域名界面为子域名添加了Worker Route，如图：
>       ![image](https://user-images.githubusercontent.com/64178177/207241812-1c6b2d08-ffec-4748-a0e9-ad64baecdab5.png)
>       ![image](https://user-images.githubusercontent.com/64178177/207243511-9919dabc-f547-40c6-8c91-7f0e8fa6875e.png)
>       （那个aa的服务就是alist-proxy的js代码）
>       此时在国内访问那个down子域名就会返回如下的Json：
>       `{ "code": 401, "message": "expire missing" }`
>       **也就是说GET请求成功转发到了workers，那个CNAME记录的添加是有效的**（大概吧）
> 
> 
> 然后我把down子域名写到了alist网盘设置的下载代理域名中，在浏览器网页点击链接并被IDM捕获之后发现下载链接的前缀从我自己的阿里云服务器域名变成了CNAME记录所写的down子域名
> 
>     * 然后这就有了第二个问题：无法下载。
>       首先，IDM捕获到下载请求之后点击下载就会报500(Internal Server Error)或者302，如图：
>       ![image](https://user-images.githubusercontent.com/64178177/207242583-6f46df25-2030-4078-943d-815ba19fbc90.png)
> 
> 
> 而用Edge下载则会根据302跳转信息跳转到那个下载页面并显示此信息： ![image](https://user-images.githubusercontent.com/64178177/207242930-767db06e-943e-440c-aa06-3716c4cadee3.png) 此时在cloudflare中查看流日志可以看到报错信息： `{ "outcome": "exception", "scriptName": "aa", "exceptions": [ { "name": "SyntaxError", "message": "Unexpected token 'e', \"error code: 1003\" is not valid JSON", "timestamp": 1670913067165 } ], "logs": [], "eventTimestamp": 1670913067160, "event": { "request": { "url": "https://down.ghngtfnjufgjnghnjfhnjfghjghjhfgjndfthdtrbhdtrh.tk/ali/MOV/%E8%80%81%E5%8F%8B%E8%AE%B0%E5%85%A8%E5%8D%81%E5%AD%A3.Friends.The.Complete.Series.1994-2003.BluRay.1080p.x265.10bit.MNHD-FRDS/%E8%80%81%E5%8F%8B%E8%AE%B0%E7%AC%AC%E5%85%AB%E5%AD%A3.Friends.S08.2001.BluRay.1080p.x265.10bit.MNHD-FRDS/Friends.S08E10.2001.BluRay.1080p.x265.10bit.MNHD-FRDS.mkv?sign=REDACTED=:0", "method": "GET", "headers": { "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9", "accept-encoding": "gzip", "accept-language": "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6", "cf-connecting-ip": "20.212.107.98", "cf-ipcountry": "SG", "cf-ray": "778ca7ad7f0e5627", "cf-visitor": "{\"scheme\":\"https\"}", "connection": "Keep-Alive", "host": "down.ghngtfnjufgjnghnjfhnjfghjghjhfgjndfthdtrbhdtrh.tk", "sec-ch-ua": "\"Not?A_Brand\";v=\"8\", \"Chromium\";v=\"108\", \"Microsoft Edge\";v=\"108\"", "sec-ch-ua-mobile": "?0", "sec-ch-ua-platform": "\"Windows\"", "sec-fetch-dest": "document", "sec-fetch-mode": "navigate", "sec-fetch-site": "cross-site", "sec-fetch-user": "?1", "upgrade-insecure-requests": "1", "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36 Edg/108.0.1462.46", "x-forwarded-proto": "https", "x-real-ip": "20.212.107.98" }, "cf": { "clientTcpRtt": 2, "longitude": "103.85540", "latitude": "1.30360", "tlsCipher": "AEAD-AES128-GCM-SHA256", "continent": "AS", "asn": 8075, "clientAcceptEncoding": "gzip, deflate, br", "country": "SG", "tlsClientAuth": { "certIssuerDNLegacy": "", "certIssuerSKI": "", "certSubjectDNRFC2253": "", "certSubjectDNLegacy": "", "certFingerprintSHA256": "", "certNotBefore": "", "certSKI": "", "certSerial": "", "certIssuerDN": "", "certVerified": "NONE", "certNotAfter": "", "certSubjectDN": "", "certPresented": "0", "certRevoked": "0", "certIssuerSerial": "", "certIssuerDNRFC2253": "", "certFingerprintSHA1": "" }, "tlsExportedAuthenticator": { "clientFinished": "395f2eeddc1c983ec34593ca5e36af1a171fcc3dc9de893bc4ac17dc745e6992", "clientHandshake": "0161a4f6c2c6c3d196d14b58ec17aeb897b643edf894c9722f4c87c0b9f6497f", "serverHandshake": "1ba578616d5b00e19664c1be6b19c782769423c7ecedfaa6d92518a86ce9221c", "serverFinished": "1489494780bbca0964b717cecec74e07020066c6d58833ae73d93ef9f5c13030" }, "tlsVersion": "TLSv1.3", "colo": "SIN", "timezone": "Asia/Singapore", "edgeRequestKeepAliveStatus": 1, "requestPriority": "weight=256;exclusive=1", "city": "Singapore", "httpProtocol": "HTTP/2", "asOrganization": "Microsoft Azure", "postalCode": "18" } }, "response": { "status": 500 } }, "id": 0 }` 看起来像是我alist的下载请求中的数据没有成功转发到workers导致的报错？这是因为我哪里的操作有问题吗？ 请问各路大神，这种问题我该如何解决？

下载会走alist服务器流量么？文件不是直接从对应网盘直接拉取么？ 我是新手

**Reply by Arc-2023** [reply]
与你的选择有关，详见(https://github.com/alist-org/alist/discussions/2279#discussioncomment-4103880)

**Comment by Arc-2023** [comment, marked as answer]
# 问题已解决：**将workers中的alist地址换成域名即可**
如果是填的ip，带端口与否、cloudflare支持端口与否cloudflare都会报1101错误

**Reply by xyooz** [reply]
你现在可以尝试这种方法，设置好worker后，进入worker→触发器→添加自定义域名，这个自定义域就是下载url。

**Reply by xyooz** [reply]
那我也不清楚了，我的步骤是，设置好域名反代→按照官方文档说的配置好worker→访问worker确认worker可用→触发器自定义域名→访问自定义域名确认可用，最后放到下载url测试，这样下来是没问题的。

**Reply by xyooz** [reply]
或许是这样，
![IMG_20230413_002530](https://user-images.githubusercontent.com/101877493/231521613-ca6dc9c3-e13f-496a-b1e5-ba8fe0466f36.jpg)
试试将域名的ssl加密模式设置为完全

**Comment by ghostWindows-files** [comment]
遇到同样的问题worker中的网盘地址加入了https：//就解决了
